File: app/ui/board.js
/* CodeBee 任务驾驶舱（/board.html）：4s 轮询 /api/board（KB 级聚合端点），
   不碰 /api/state 全量，也不占 SSE。 */
"use strict";

const $ = (s) => document.querySelector(s);
const t = (k) => (window.t ? window.t(k) : k);   // i18n.js 先于本脚本加载
const POLL_MS = 4000;

// URL ?lang=en|zh 可覆盖语言（i18n.js 已按 localStorage 自动 apply 过一次，
// 这里写回偏好并重刷静态文案；动态文案走 t() 每次渲染实时取）
try {
  const ql = new URLSearchParams(location.search).get("lang");
  if ((ql === "en" || ql === "zh") && window.getLang && window.getLang() !== ql) {
    window.setLang(ql);
    if (window.applyI18n) window.applyI18n();
  }
} catch (e) { /* ignore */ }

/* ---------------------------------------------------------------- 工具 */
function fmtTokens(n) {
  n = Number(n) || 0;
  if (n >= 1e8) return (n / 1e8).toFixed(2) + t("亿");
  if (n >= 1e4) return (n / 1e4).toFixed(1) + t("万");
  if (n >= 1e3) return (n / 1e3).toFixed(1) + "k";
  return String(n);
}
function fmtCost(n) {
  n = Number(n) || 0;
  if (n >= 100) return "$" + n.toFixed(0);
  if (n >= 1) return "$" + n.toFixed(2);
  return "$" + n.toFixed(3);
}
function fmtDur(sec) {
  sec = Math.max(0, Math.floor(Number(sec) || 0));
  const h = Math.floor(sec / 3600), m = Math.floor((sec % 3600) / 60), s = sec % 60;
  if (h) return h + "h" + String(m).padStart(2, "0") + "m";
  return String(m).padStart(2, "0") + ":" + String(s).padStart(2, "0");
}
function parseTs(s) {
  // 后端时间格式 "YYYY-MM-DD HH:MM:SS"；补 T 让各浏览器 Date 都能解析
  const t = Date.parse(String(s || "").replace(" ", "T"));
  return isNaN(t) ? 0 : t;
}
function esc(s) {
  return String(s == null ? "" : s).replace(/[&<>"']/g,
    (c) => ({"&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;"}[c]));
}

/* ---------------------------------------------------------------- 令牌门 */
function token() { return localStorage.getItem("orch.token") || ""; }

function showGate(msg) {
  $("#gate").classList.remove("hidden");
  $("#gate-err").textContent = msg || "";
  setTimeout(() => $("#gate-token").focus(), 50);
}
$("#gate-ok").addEventListener("click", submitGate);
$("#gate-token").addEventListener("keydown", (e) => { if (e.key === "Enter") submitGate(); });
function submitGate() {
  const v = $("#gate-token").value.trim();
  if (!v) return;
  localStorage.setItem("orch.token", v);
  $("#gate").classList.add("hidden");
  poll();
}

/* ---------------------------------------------------------------- 拉取 */
let connLost = false;

async function poll() {
  let data;
  try {
    const r = await fetch("/api/board", { headers: { "X-CodeBee-Token": token() } });
    if (r.status === 401) { showGate(t("令牌不正确或已变更")); return; }
    if (!r.ok) throw new Error("HTTP " + r.status);
    data = await r.json();
    if (connLost) { connLost = false; $("#hd-conn").classList.add("hidden"); }
  } catch (e) {
    if (!connLost) { connLost = true; $("#hd-conn").classList.remove("hidden"); }
    return;
  }
  render(data);
}

/* ---------------------------------------------------------------- 渲染 */
const cardEls = new Map();   // task_id -> 卡元素（复用防闪烁）

function render(d) {
  // KPI
  $("#k-running").textContent = d.running.length;
  $("#k-queued").textContent = d.queued.length;
  $("#k-done").textContent = d.today_done;
  $("#k-failed").textContent = d.today_failed;
  $("#k-tokens").textContent = fmtTokens(d.usage_today.tokens);
  $("#k-cost").textContent = fmtCost(d.usage_today.cost_usd);

  // 顶栏
  const pool = d.pool || {};
  $("#hd-pool").textContent = t("蜂巢") + " " + (pool.alive || 0) + "/" + (pool.target || 0)
    + " · " + t("对话") + " " + (pool.chat_alive || 0) + "/" + (pool.chat_pool || 0);
  const downs = (d.health || []).filter((h) => h.status === "down" && !h.silenced);
  const hp = $("#hd-health");
  if (downs.length) {
    hp.classList.remove("hidden");
    hp.textContent = "⚠ " + downs.map((h) => h.provider).join("、") + " " + t("异常");
    hp.title = downs.map((h) => (h.provider + " " + (h.model || "")).trim()).join("\n");
  } else {
    hp.classList.add("hidden");
  }

  renderWall(d);
  renderTrend(d.trend || []);
  renderModels(d.models || []);
  renderRecent(d.recent || []);
}

function renderWall(d) {
  const wrap = $("#wall-cards");
  const empty = $("#wall-empty");
  const has = d.running.length > 0;
  empty.classList.toggle("hidden", has);
  wrap.classList.toggle("hidden", !has);

  const seen = new Set();
  for (const c of d.running) {
    seen.add(c.task_id);
    let el = cardEls.get(c.task_id);
    if (!el) {
      el = document.createElement("div");
      el.className = "card";
      el.innerHTML =
        '<div class="card-top"><span class="c-type"></span><span class="c-title"></span>' +
        '<span class="c-elapsed" data-started=""></span></div>' +
        '<div class="c-meta"><span class="b b-model"></span><span class="b b-prov"></span>' +
        '<span class="b b-agent"></span><span class="c-run-tok"></span></div>' +
        '<div class="c-steps"></div>' +
        '<div class="c-stepinfo"><span class="no"></span><span class="ag"></span><span class="sm"></span></div>' +
        '<div class="c-live"><span class="c-live-tag"></span><pre class="c-live-text"></pre></div>' +
        '<div class="c-pct"></div>';
      wrap.appendChild(el);
      cardEls.set(c.task_id, el);
    }
    el.querySelector(".c-type").textContent = c.type_name || c.type || t("任务");
    el.querySelector(".c-title").textContent = c.title || t("（未命名）");
    const el2 = el.querySelector(".c-elapsed");
    el2.dataset.started = c.started_at || "";
    el2.textContent = fmtDur((Date.now() - parseTs(c.started_at)) / 1000);

    // 模型 / 厂商 / 执行智能体徽章 + 本次消耗
    const cur0 = c.current || {};
    const bModel = el.querySelector(".b-model");
    bModel.textContent = cur0.model || t("模型待定");
    bModel.classList.toggle("dim", !cur0.model);
    const bProv = el.querySelector(".b-prov");
    bProv.textContent = cur0.provider || "";
    bProv.classList.toggle("hidden", !cur0.provider);
    const bAgent = el.querySelector(".b-agent");
    bAgent.textContent = cur0.agent || "";
    bAgent.classList.toggle("hidden", !cur0.agent);
    const rt = el.querySelector(".c-run-tok");
    const cost = Number(c.run_cost_usd) || 0;
    rt.textContent = (c.run_tokens ? fmtTokens(c.run_tokens) + " tok" : "") +
      (cost > 0 ? (c.run_tokens ? " · " : "") + "$" + cost.toFixed(cost >= 1 ? 2 : 3) : "");
    rt.classList.toggle("hidden", !c.run_tokens && !(cost > 0));

    // 分段进度条
    const seg = el.querySelector(".c-steps");
    const total = Math.max(c.steps_total || 0, 1);
    if (seg.childElementCount !== total) {
      seg.innerHTML = "";
      for (let i = 0; i < total; i++) seg.appendChild(document.createElement("i"));
    }
    [...seg.children].forEach((b, i) => {
      b.className = i < (c.steps_done || 0) ? "d" : (i === (c.steps_done || 0) ? "r" : "");
    });

    const cur = c.current || {};
    el.querySelector(".no").textContent = cur.n ? ("#" + cur.n) : "—";
    el.querySelector(".ag").textContent = cur.agent || "";
    el.querySelector(".sm").textContent = cur.summary || t("执行中");

    // 实时流窗口：思考/输出尾部多行（live_label 是 code，前端翻译）
    const live = el.querySelector(".c-live");
    const lt = cur.live_text || "";
    live.classList.toggle("hidden", !lt);
    if (lt) {
      const tagMap = { output: t("输出"), thinking: t("思考") };
      el.querySelector(".c-live-tag").textContent = tagMap[cur.live_label] || t("实时");
      el.querySelector(".c-live-text").textContent = lt;
    }

    // 有预估时长则给完成度百分比
    const pct = el.querySelector(".c-pct");
    if (c.estimate_s && parseTs(c.started_at)) {
      const p = Math.min(99, Math.round(((Date.now() - parseTs(c.started_at)) / 1000) / c.estimate_s * 100));
      pct.textContent = t("预估 {0} · 进度≈{1}%", fmtDur(c.estimate_s), p);
    } else {
      pct.textContent = "";
    }
  }
  for (const [tid, el] of cardEls) {
    if (!seen.has(tid)) { el.remove(); cardEls.delete(tid); }
  }

  // 排队条
  const qr = $("#queue-row");
  qr.classList.toggle("hidden", !d.queued.length);
  $("#queue-chips").innerHTML = d.queued.map((q) =>
    '<span class="chip" title="' + esc(q.title) + '">' + esc(q.title) + "</span>").join("");
}

function renderTrend(trend) {
  const box = $("#trend");
  const max = Math.max(1, ...trend.map((t) => t.tokens || 0));
  box.innerHTML = trend.map((t) => {
    const h = Math.round((t.tokens || 0) / max * 100);
    const label = String(t.day || "").slice(5);   // MM-DD
    return '<div class="bar' + (t.tokens ? "" : " zero") + '" title="' + esc(t.day) + " · " +
      fmtTokens(t.tokens) + '"><i style="height:' + Math.max(h, 3) + '%"></i><b>' + esc(label) + "</b></div>";
  }).join("");
}

function renderModels(models) {
  const box = $("#models");
  if (!models.length) { box.innerHTML = '<div class="empty">' + t("暂无调用记录") + "</div>"; return; }
  const max = Math.max(1, ...models.map((m) => m.tokens || 0));
  box.innerHTML = models.map((m) =>
    '<div class="model-row"><span class="nm" title="' + esc(m.model) + '">' + esc(m.model) + "</span>" +
    '<span class="tk"><i style="width:' + Math.max(4, Math.round((m.tokens || 0) / max * 100)) + '%"></i></span>' +
    '<span class="vl">' + fmtTokens(m.tokens) + "</span></div>").join("");
}

function renderRecent(recent) {
  const box = $("#recent");
  if (!recent.length) { box.innerHTML = '<div class="empty">' + t("还没有已完成的运行") + "</div>"; return; }
  const dot = { done: "done", failed: "failed", cancelled: "cancelled" };
  box.innerHTML = recent.map((r) => {
    const cls = dot[r.status] || "cancelled";
    const t = String(r.ended_at || "").slice(11, 16);
    return '<div class="r-row"><span class="r-dot ' + cls + '"></span>' +
      '<span class="r-t' + (r.status === "failed" ? " failed" : "") + '" title="' +
      esc(r.title + (r.error ? " · " + r.error : "")) + '">' + esc(r.title) + "</span>" +
      '<span class="r-d">' + (r.duration_s != null ? fmtDur(r.duration_s) : "—") + " · " + esc(t) + "</span></div>";
  }).join("");
}

/* ---------------------------------------------------------------- 时钟与耗时每秒跳 */
setInterval(() => {
  const n = new Date();
  const p = (x) => String(x).padStart(2, "0");
  $("#hd-clock").textContent = p(n.getHours()) + ":" + p(n.getMinutes()) + ":" + p(n.getSeconds());
  document.querySelectorAll(".c-elapsed[data-started]").forEach((el) => {
    const st = parseTs(el.dataset.started);
    if (st) el.textContent = fmtDur((Date.now() - st) / 1000);
  });
}, 1000);

poll();
setInterval(poll, POLL_MS);

File: app/core/board.py
# -*- coding: utf-8 -*-
"""任务驾驶舱（/board 大屏）的轻量聚合数据源。

大屏 4 秒轮询：不能拉 /api/state 全量（MB 级），也不能每次全量扫 usage
台账（summary 全历史读入），所以这里只取大屏要看的字段，并把 usage 聚合
缓存 8 秒。只读 store/usage/jobs/health，不落新盘。
"""
from __future__ import annotations

import time

_CACHE = {"ts": 0.0, "data": None}
_TTL = 8.0

_TAIL_MAX = 150          # 跑马灯尾行截断
_TAIL_LINES = 6          # 实时流窗口行数
_TAIL_LINE_CHARS = 56    # 实时流单行截断（卡内 11px 等宽不换行的安全宽度）
_RECENT_LIMIT = 10       # 最近完成/失败条数
_RUN_WINDOW = 300        # 今日计数扫描的运行窗口（id 字典序=时间序）


def payload():
    now = time.time()
    if _CACHE["data"] is not None and now - _CACHE["ts"] < _TTL:
        return _CACHE["data"]
    data = _build()
    _CACHE["ts"] = now
    _CACHE["data"] = data
    return data


def _tail(step):
    """步骤实时流（正文优先，其次思考）的最后一行，给大屏跑马灯。"""
    for key in ("stream", "thinking"):
        v = step.get(key)
        if isinstance(v, str) and v.strip():
            lines = [ln.strip() for ln in v.strip().splitlines() if ln.strip()]
            if lines:
                return lines[-1][:_TAIL_MAX]
    return ""


def _live_block(step):
    """步骤实时流窗口：尾部多行（正文优先，其次思考），供大屏看它在想什么。

    返回 (来源标签 code, 文本)；标签是 code（output/thinking），由前端 i18n
    翻译——大屏要做多语言，后端不下发中文。都没有返回 ("", "")。
    """
    for key, code in (("stream", "output"), ("thinking", "thinking")):
        v = step.get(key)
        if not (isinstance(v, str) and v.strip()):
            continue
        lines = [ln.strip()[:_TAIL_LINE_CHARS] for ln in v.strip().splitlines()]
        lines = [ln for ln in lines if ln]
        if lines:
            return code, "\n".join(lines[-_TAIL_LINES:])
    return "", ""


def _model_provider_map():
    """模型名 → 厂商名（enabled 优先，同模型多厂商取 priority 最靠前的）。"""
    try:
        from . import modelhub
        rows = modelhub.models_view()
    except Exception:
        return {}
    best = {}
    for r in rows:
        name = r.get("name")
        if not name or not r.get("enabled", False):
            continue
        pri = r.get("priority", 0)
        cur = best.get(name)
        if cur is None or pri < cur[0]:
            best[name] = (pri, r.get("provider_name") or r.get("provider_id") or "")
    return {k: v[1] for k, v in best.items()}


def _parse_ts(s):
    try:
        return time.mktime(time.strptime(str(s), "%Y-%m-%d %H:%M:%S"))
    except (ValueError, TypeError):
        return 0.0


def _duration_s(run):
    """运行耗时（秒）：ended-started，缺 started 用 created 兜底。"""
    end = _parse_ts(run.get("ended_at"))
    start = _parse_ts(run.get("started_at")) or _parse_ts(run.get("created_at"))
    if end > 0 and start > 0 and end >= start:
        return int(end - start)
    return None


def _step_view(steps, prov_map):
    """进度概要：完成数 + 当前运行中步骤（含模型/厂商/实时流）。"""
    done = sum(1 for s in steps if s.get("status") == "done")
    cur = next((s for s in steps if s.get("status") == "running"), None)
    cur_v = None
    if cur:
        live_label, live_text = _live_block(cur)
        model = str(cur.get("model") or "")[:60]
        cur_v = {"n": cur.get("n"), "role": cur.get("role") or "",
                 "agent": cur.get("agent_label") or cur.get("agent") or "",
                 "summary": (cur.get("summary") or "")[:80],
                 "tail": _tail(cur),
                 "model": model, "provider": prov_map.get(model, "") if model else "",
                 "live_label": live_label, "live_text": live_text}
    return done, len(steps), cur_v


def _build():
    from . import flows, health, jobs, store, usage

    flow_map = {}
    try:
        for f in flows.list_flows():
            flow_map[f.get("id")] = f
    except Exception:
        pass

    prov_map = _model_provider_map()
    today = time.strftime("%Y-%m-%d")
    tasks = store.list_tasks(200, archived=False)
    latest = store.latest_run_by_task()

    running, queued = [], []
    for t in tasks:
        tid = t.get("id")
        run = latest.get(tid)
        status = t.get("status")
        if status == "running" and run and run.get("status") == "running":
            done, total, cur = _step_view(run.get("steps") or [], prov_map)
            running.append({
                "task_id": tid, "run_id": run.get("id"),
                "title": t.get("title") or "", "goal": (t.get("goal") or "")[:120],
                "type": t.get("type") or "",
                "type_name": (flow_map.get(t.get("type")) or {}).get("name") or (t.get("type") or ""),
                "status": status,
                "started_at": run.get("started_at") or run.get("created_at") or "",
                "steps_done": done, "steps_total": total,
                "current": cur,
                "run_tokens": run.get("tokens") or 0,
                "run_cost_usd": round(float(run.get("cost_usd") or 0.0), 4),
                "estimate_s": run.get("estimated_duration_s"),
            })
        elif status == "queued":
            queued.append({
                "task_id": tid, "run_id": (run or {}).get("id"),
                "title": t.get("title") or "",
                "type_name": (flow_map.get(t.get("type")) or {}).get("name") or (t.get("type") or ""),
                "created_at": (run or {}).get("created_at") or "",
            })
    running.sort(key=lambda c: _parse_ts(c["started_at"]))
    queued.sort(key=lambda c: _parse_ts(c["created_at"]))

    runs = store.list_runs(_RUN_WINDOW)
    today_done = today_failed = 0
    recent = []
    for r in runs:
        st = r.get("status")
        ended = str(r.get("ended_at") or "")
        if st in ("done", "failed", "cancelled") and ended.startswith(today):
            if st == "done":
                today_done += 1
            elif st == "failed":
                today_failed += 1
        if len(recent) < _RECENT_LIMIT and st in ("done", "failed", "cancelled") and ended:
            recent.append({
                "run_id": r.get("id"), "task_id": r.get("task_id"),
                "title": r.get("title") or "", "status": st,
                "ended_at": ended, "duration_s": _duration_s(r),
                "error": (r.get("error") or "")[:80],
            })

    trend, models, usage_today = [], [], {"tokens": 0, "calls": 0, "cost_usd": 0.0}
    try:
        s = usage.summary(days=7)
        trend = [{"day": d.get("day"), "tokens": d.get("tokens") or 0}
                 for d in (s.get("by_day") or [])]
        for row in (s.get("by_model") or [])[:5]:
            if not row.get("tokens"):
                continue
            models.append({"model": row.get("key") or "?",
                           "tokens": row.get("tokens") or 0,
                           "calls": row.get("calls") or 0})
        for d in s.get("by_day") or []:
            if str(d.get("day")) == today:
                usage_today = {"tokens": d.get("tokens") or 0,
                               "calls": d.get("calls") or 0,
                               "cost_usd": round(float(d.get("cost_usd") or 0.0), 3)}
                break
    except Exception:
        pass

    return {
        "now": time.strftime("%Y-%m-%d %H:%M:%S"),
        "today": today,
        "running": running,
        "queued": queued,
        "today_done": today_done,
        "today_failed": today_failed,
        "recent": recent,
        "trend": trend,
        "models": models,
        "usage_today": usage_today,
        "pool": jobs.workers_info(),
        "health": health.snapshot().get("providers") or [],
    }
